evm_loader/program/src/account/sysvar.rs

```diff
@@ -0,0 +1,143 @@
+    /// Returns the pair of (`compute_budget_unit_limit`, `compute_budget_unit_price`)
+    /// N.B. the `compute_budget_unit_price` is denominated in micro Lamports.
+    pub fn get_compute_budget_priority_fee(&self) -> Result<(u32, u64), Error> {
+        let compute_budget_account_pubkey =
+            Pubkey::from_str(COMPUTE_BUDGET_ADDRESS).map_err(|_| {
+                Error::PriorityFeeParsingError("Invalid Compute budget address.".to_string())
+            })?;
+        // Intent is to check all the instructions before the current one.
+        let max_idx = instructions::load_current_index_checked(self.0)? as usize;
+
+        let mut idx = 0;
+        let mut compute_unit_limit: Option<u32> = None;
+        let mut compute_unit_price: Option<u64> = None;
+        while (compute_unit_limit.is_none() || compute_unit_price.is_none()) && idx < max_idx {
+            let cur_ixn = instructions::load_instruction_at_checked(idx, self.0)?;
+
+            // Skip all instructions that do not target Compute Budget Program.
+            if cur_ixn.program_id != compute_budget_account_pubkey {
+                idx += 1;
+                continue;
+            }
+
+            // As of now, data of ComputeBudgetInstruction is always non-empty.
+            // This is a sanity check to have a safe future-proof implementation.
+            let tag = cur_ixn.data.first().unwrap_or(&0);
+            match *tag {
+                COMPUTE_UNIT_LIMIT_TAG => {
+                    compute_unit_limit = Some(u32::from_le_bytes(
+                        cur_ixn.data[1..].try_into().map_err(|_| {
+                            Error::PriorityFeeParsingError(
+                                "Invalid format of compute unit limit.".to_string(),
+                            )
+                        })?,
+                    ));
+                }
+                COMPUTE_UNIT_PRICE_TAG => {
+                    compute_unit_price = Some(u64::from_le_bytes(
+                        cur_ixn.data[1..].try_into().map_err(|_| {
+                            Error::PriorityFeeParsingError(
+                                "Invalid format of compute unit price.".to_string(),
+                            )
+                        })?,
+                    ));
+                }
+                _ => (),
+            }
+            idx += 1;
+        }
+
+        if compute_unit_price.is_none() {
+            return Err(Error::PriorityFeeNotSpecified);
+        }
+
+        // Caller may not specify the compute unit limit, the default should take effect.
+        if compute_unit_limit.is_none() {
+            compute_unit_limit = Some(DEFAULT_COMPUTE_UNIT_LIMIT);
+        }
+
+        // Both are not none, it's safe to unwrap.
+        Ok((compute_unit_limit.unwrap(), compute_unit_price.unwrap()))
+    }
+
+    /// Checks that priority fee as set by the Operator is accurate to what User set as `max_priority_fee_per_gas`.
```

Comment:
Should we check if the block's fullness so we can validate if the supplied fee follows the expected multiplier curve i.e [-12.5%, 12.5%], or do we leave this up to the user?

Reply:
Short answer: no, it's not possible in the current architecture.

Longer answer: There's no way to check the "blockchain" state on the EVM side. EVM simply does not know about the existence of blocks that indexer assembes.

So the decision was to make base_fee_per_gas **NOT** follow the Ethereum semantics (to have it float from -12% to +12% based off of prev block fullness).
Instead, base_fee_per_gas for new transaction is exactly the same as current gas_price for legacy transactions: it includes token_price_multiplier (SOL/gasToken price ratio), slippage multiplier and operator's profit multiplier.

Hope it makes sense.